File: 1st/homework_10/id_45/120_triangle.cpp
class Solution {
public:
    int minimumTotal(vector<vector<int>>& triangle) {
        int n = triangle.size();

        if (n == 0)
            return 0;

        if (n == 1)
            return triangle[0][0];

        vector<int> cur_min_sum = triangle[n - 1]; 
        for (int i = n - 2; i >= 0; --i) {
            for (int j = 0; j < triangle[i].size(); ++j) {
                cur_min_sum[j] = triangle[i][j] + min(cur_min_sum[j], cur_min_sum[j+1]);
            }
        }   
        return cur_min_sum[0];
    }
};

File: 4th/homework_6/id_14/120_Triangle.py
'''
Given a triangle, find the minimum path sum from top to bottom.
Each step you may move to adjacent numbers on the row below.

For example, given the following triangle
[
     [2],
    [3,4],
   [6,5,7],
  [4,1,8,3]
]
The minimum path sum from top to bottom is 11 (i.e., 2 + 3 + 5 + 1 = 11).

Note:
Bonus point if you are able to do this using only O(n) extra space, where n is the total number of rows in the triangle.
'''

# 自下而上
class Solution(object):
    def minimumTotal(self, triangle):
        """
        :type triangle: List[List[int]]
        :rtype: int
        """
        if not triangle:
            return 0
        ret = triangle[-1][:]
        for i in range(len(triangle) - 2, -1, -1):
            for j in range(len(triangle[i])):
                ret[j] = triangle[i][j] + min(ret[j], ret[j + 1])
        return ret[0]

# triangle = [[-1],[2,3],[1,-1,-3]]
triangle = [
     [2],
    [3,4],
   [6,5,7],
  [4,1,8,3]
]
s = Solution()
print(s.minimumTotal(triangle))
print(triangle)